// @progress/kendo-angular-grid/dist/es/table-body.component.js
import { Component, Input } from '@angular/core';
import { DetailsService } from './details.service';
import { GroupsService } from './grouping/groups.service';
import { ChangeNotificationService } from './change-notification.service';
import { isChanged } from './utils';
/**
 * @hidden
 */
export var TableBodyComponent = (function () {
    function TableBodyComponent(detailsService, groupsService, changeNotification) {
        this.detailsService = detailsService;
        this.groupsService = groupsService;
        this.changeNotification = changeNotification;
        this.columns = [];
        this.groups = [];
        this.skip = 0;
        this.skipGroupDecoration = false;
        this.showGroupFooters = false;
    }
    TableBodyComponent.prototype.toggleRow = function (index, dataItem) {
        this.detailsService.toggleRow(index, dataItem);
        return false;
    };
    TableBodyComponent.prototype.trackByFn = function (_, item) {
        return item.data ? item.data : item;
    };
    TableBodyComponent.prototype.isExpanded = function (index) {
        return this.detailsService.isExpanded(index);
    };
    TableBodyComponent.prototype.detailButtonStyles = function (index) {
        var expanded = this.isExpanded(index);
        return { 'k-minus': expanded, 'k-plus': !expanded };
    };
    TableBodyComponent.prototype.isGroup = function (item) {
        return item.type === 'group';
    };
    TableBodyComponent.prototype.isDataItem = function (item) {
        return !this.isGroup(item) && !this.isFooter(item);
    };
    TableBodyComponent.prototype.isFooter = function (item) {
        return item.type === 'footer';
    };
    TableBodyComponent.prototype.isInExpandedGroup = function (item) {
        return this.groupsService.isInExpandedGroup(item.groupIndex, false);
    };
    TableBodyComponent.prototype.isParentGroupExpanded = function (item) {
        return this.groupsService.isInExpandedGroup(item.index || item.groupIndex);
    };
    TableBodyComponent.prototype.ngOnChanges = function (changes) {
        if (isChanged("columns", changes, false)) {
            this.changeNotification.notify();
        }
    };
    TableBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: '[kendoGridTableBody]',
                    template: "\n    <template ngFor\n        [ngForOf]=\"data\"\n        [ngForTrackBy]=\"trackByFn\"\n        let-item\n        let-odd=\"odd\">\n        <tr *ngIf=\"isGroup(item) && isParentGroupExpanded(item)\"\n            kendoGridGroupHeader\n            [columns]=\"columns\"\n            [groups]=\"groups\"\n            [item]=\"item\"\n            [skipGroupDecoration]=\"skipGroupDecoration\"\n            >\n        </tr>\n        <tr\n            *ngIf=\"isDataItem(item) && isInExpandedGroup(item)\"\n            [class.k-alt]=\"odd\"\n            [class.k-master-row]=\"detailTemplate?.templateRef\"\n            [kendoGridSelectable]=\"selectable\"\n            [index]=\"item.index\">\n            <template [ngIf]=\"!skipGroupDecoration\">\n                <td [class.k-group-cell]=\"true\" *ngFor=\"let g of groups\"></td>\n            </template>\n            <td [class.k-hierarchy-cell]=\"true\"\n                *ngIf=\"detailTemplate?.templateRef\">\n                <a class=\"k-icon\"\n                    [ngClass]=\"detailButtonStyles(item.index)\"\n                    href=\"#\" tabindex=\"-1\" (click)=\"toggleRow(item.index, item.data)\"></a>\n            </td>\n            <td\n                [ngClass]=\"column.cssClass\"\n                [ngStyle]=\"column.style\"\n                *ngFor=\"let column of columns; let columnIndex = index\">\n                <template\n                    [templateContext]=\"{\n                        templateRef: column.templateRef,\n                        dataItem: item.data,\n                        column: column,\n                        columnIndex: columnIndex,\n                        rowIndex: item.index,\n                        $implicit: item.data\n                        }\">\n                </template>\n                <template [ngIf]=\"!column.templateRef\">{{item.data | valueOf: column.field: column.format}}</template>\n            </td>\n        </tr>\n        <tr *ngIf=\"isDataItem(item) && detailTemplate?.templateRef && isExpanded(item.index)\"\n            [class.k-detail-row]=\"true\"\n            [class.k-alt]=\"odd\">\n            <td [class.k-group-cell]=\"true\" *ngFor=\"let g of groups\"></td>\n            <td [class.k-hierarchy-cell]=\"true\"></td>\n            <td [class.k-detail-cell]=\"true\"\n                [attr.colspan]=\"columns.length\">\n                <template\n                    [templateContext]=\"{\n                        templateRef: detailTemplate?.templateRef,\n                        dataItem: item.data,\n                        rowIndex: item.index,\n                        $implicit: item.data\n                        }\">\n                </template>\n            </td>\n        </tr>\n        <tr *ngIf=\"isFooter(item) && (isInExpandedGroup(item) || (showGroupFooters && isParentGroupExpanded(item)))\" \n            [class.k-group-footer]=\"true\">\n            <template [ngIf]=\"!skipGroupDecoration\">\n                <td [class.k-group-cell]=\"true\" *ngFor=\"let g of groups\"></td>\n            </template>\n            <td [class.k-hierarchy-cell]=\"true\"\n                *ngIf=\"detailTemplate?.templateRef\">\n            </td>\n            <td\n                *ngFor=\"let column of columns; let columnIndex = index\">\n                <template\n                    [templateContext]=\"{\n                        templateRef: column.groupFooterTemplateRef,\n                        group: item.data,\n                        field: column.field,\n                        column: column,\n                        $implicit: item.data?.aggregates\n                        }\">\n                </template>\n           </td>\n        </tr>\n    </template>\n    "
                },] },
    ];
    /** @nocollapse */
    TableBodyComponent.ctorParameters = [
        { type: DetailsService, },
        { type: GroupsService, },
        { type: ChangeNotificationService, },
    ];
    TableBodyComponent.propDecorators = {
        'columns': [{ type: Input },],
        'groups': [{ type: Input },],
        'detailTemplate': [{ type: Input },],
        'data': [{ type: Input },],
        'skip': [{ type: Input },],
        'selectable': [{ type: Input },],
        'skipGroupDecoration': [{ type: Input },],
        'showGroupFooters': [{ type: Input },],
    };
    return TableBodyComponent;
}());
