// angular2localization/src/directives/locale-number-validator.directive.js
/**
 * ANGULAR 2 LOCALIZATION
 * An Angular 2 library to translate messages, dates and numbers.
 * Written by Roberto Simonetti.
 * MIT license.
 * https://github.com/robisim74/angular2localization
 */
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { LocaleService } from '../services/locale.service';
import { LocaleParser } from '../services/locale-parser';
/**
 * Function that takes a Control and returns either null when it’s valid, or and error object if it’s not.

\@param locale The reference to LocaleService
\@param digits The format of the number
\@param MIN_VALUE The minimum value for the number
\@param MAX_VALUE The maximum value for the number
\@return An error object: 'format', 'minValue' or 'maxValue'; null in case the value is valid
 * @param {?} locale
 * @param {?} digits
 * @param {?=} MIN_VALUE
 * @param {?=} MAX_VALUE
 * @return {?}
 */
export function validateLocaleNumber(locale, digits, MIN_VALUE, MAX_VALUE) {
    if (MIN_VALUE === void 0) { MIN_VALUE = Number.MIN_VALUE; }
    if (MAX_VALUE === void 0) { MAX_VALUE = Number.MAX_VALUE; }
    var /** @type {?} */ defaultLocale;
    var /** @type {?} */ NUMBER_REGEXP;
    return function (c) {
        // Checks if the default locale has changed. 
        if (defaultLocale != locale.getDefaultLocale()) {
            NUMBER_REGEXP = LocaleParser.NumberRegExpFactory(locale.getDefaultLocale(), digits);
            defaultLocale = locale.getDefaultLocale();
        }
        // Checks the format.
        if (NUMBER_REGEXP.test(c.value)) {
            var /** @type {?} */ parsedValue;
            parsedValue = LocaleParser.Number(c.value, locale.getDefaultLocale());
            if (parsedValue < MIN_VALUE) {
                return {
                    minValue: {
                        valid: false
                    }
                };
            }
            else if (parsedValue > MAX_VALUE) {
                return {
                    maxValue: {
                        valid: false
                    }
                };
            }
            return null; // The number is valid.
        }
        else {
            return {
                format: {
                    valid: false
                }
            };
        }
    };
}
/**
 * LocaleNumberValidator class.
Validates a number by default locale.

\@author Roberto Simonetti
 */
export var LocaleNumberValidator = (function () {
    /**
     * @param {?} locale
     */
    function LocaleNumberValidator(locale) {
        this.locale = locale;
        this.MIN_VALUE = Number.MIN_VALUE;
        this.MAX_VALUE = Number.MAX_VALUE;
    }
    Object.defineProperty(LocaleNumberValidator.prototype, "minValue", {
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.MIN_VALUE = value || this.MIN_VALUE;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LocaleNumberValidator.prototype, "maxValue", {
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.MAX_VALUE = value || this.MAX_VALUE;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    LocaleNumberValidator.prototype.ngOnInit = function () {
        this.validator = validateLocaleNumber(this.locale, this.digits, this.MIN_VALUE, this.MAX_VALUE);
    };
    /**
     * @param {?} c
     * @return {?}
     */
    LocaleNumberValidator.prototype.validate = function (c) {
        return this.validator(c);
    };
    LocaleNumberValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[validateLocaleNumber][ngModel],[validateLocaleNumber][formControl]',
                    providers: [
                        { provide: NG_VALIDATORS, useExisting: forwardRef(function () { return LocaleNumberValidator; }), multi: true }
                    ]
                },] },
    ];
    /** @nocollapse */
    LocaleNumberValidator.ctorParameters = function () { return [
        { type: LocaleService, },
    ]; };
    LocaleNumberValidator.propDecorators = {
        'digits': [{ type: Input, args: ['validateLocaleNumber',] },],
        'minValue': [{ type: Input },],
        'maxValue': [{ type: Input },],
    };
    return LocaleNumberValidator;
}());
function LocaleNumberValidator_tsickle_Closure_declarations() {
    /** @type {?} */
    LocaleNumberValidator.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    LocaleNumberValidator.ctorParameters;
    /** @type {?} */
    LocaleNumberValidator.propDecorators;
    /**
     * Format: {minIntegerDigits}.{minFractionDigits}-{maxFractionDigits}.
     * @type {?}
     */
    LocaleNumberValidator.prototype.digits;
    /** @type {?} */
    LocaleNumberValidator.prototype.MIN_VALUE;
    /** @type {?} */
    LocaleNumberValidator.prototype.MAX_VALUE;
    /** @type {?} */
    LocaleNumberValidator.prototype.validator;
    /** @type {?} */
    LocaleNumberValidator.prototype.locale;
}
