// angular2localization/angular2localization.js
// Exports services, pipes & directives.
export { LocalizationService, ServiceState, LoadingMode } from './src/services/localization.service';
export { LocaleService } from './src/services/locale.service';
export { Locale } from './src/services/locale';
export { IntlSupport } from './src/services/Intl-support';
export { LocaleParser, NumberCode } from './src/services/locale-parser';
export { LocaleDirective, I18nPlural } from './src/services/i18n';
export { TranslatePipe } from './src/pipes/translate.pipe';
export { LocaleDatePipe } from './src/pipes/locale-date.pipe';
export { LocaleDecimalPipe, LocalePercentPipe, LocaleCurrencyPipe } from './src/pipes/locale-number.pipe';
export { TranslateDirective } from './src/directives/translate.directive';
export { LocaleDateDirective } from './src/directives/locale-date.directive';
export { LocaleDecimalDirective, LocalePercentDirective, LocaleCurrencyDirective } from './src/directives/locale-number.directive';
export { LocaleNumberValidator, validateLocaleNumber } from './src/directives/locale-number-validator.directive';
import { NgModule } from '@angular/core';
import { LocaleService } from './src/services/locale.service';
import { LocalizationService } from './src/services/localization.service';
import { TranslatePipe } from './src/pipes/translate.pipe';
import { LocaleDatePipe } from './src/pipes/locale-date.pipe';
import { LocaleDecimalPipe, LocalePercentPipe, LocaleCurrencyPipe } from './src/pipes/locale-number.pipe';
import { TranslateDirective } from './src/directives/translate.directive';
import { LocaleDateDirective } from './src/directives/locale-date.directive';
import { LocaleDecimalDirective, LocalePercentDirective, LocaleCurrencyDirective } from './src/directives/locale-number.directive';
import { LocaleNumberValidator } from './src/directives/locale-number-validator.directive';
export var LocaleModule = (function () {
    function LocaleModule() {
    }
    /**
     * @return {?}
     */
    LocaleModule.forRoot = function () {
        return {
            ngModule: LocaleModule,
            providers: [LocaleService]
        };
    };
    /**
     * @return {?}
     */
    LocaleModule.forChild = function () {
        return {
            ngModule: LocaleModule,
            providers: [LocaleService]
        };
    };
    LocaleModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        LocaleDatePipe,
                        LocaleDecimalPipe,
                        LocalePercentPipe,
                        LocaleCurrencyPipe,
                        LocaleDateDirective,
                        LocaleDecimalDirective,
                        LocalePercentDirective,
                        LocaleCurrencyDirective,
                        LocaleNumberValidator
                    ],
                    exports: [
                        LocaleDatePipe,
                        LocaleDecimalPipe,
                        LocalePercentPipe,
                        LocaleCurrencyPipe,
                        LocaleDateDirective,
                        LocaleDecimalDirective,
                        LocalePercentDirective,
                        LocaleCurrencyDirective,
                        LocaleNumberValidator
                    ]
                },] },
    ];
    /** @nocollapse */
    LocaleModule.ctorParameters = function () { return []; };
    return LocaleModule;
}());
function LocaleModule_tsickle_Closure_declarations() {
    /** @type {?} */
    LocaleModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    LocaleModule.ctorParameters;
}
export var LocalizationModule = (function () {
    function LocalizationModule() {
    }
    /**
     * @return {?}
     */
    LocalizationModule.forRoot = function () {
        return {
            ngModule: LocalizationModule,
            providers: [LocalizationService]
        };
    };
    /**
     * @return {?}
     */
    LocalizationModule.forChild = function () {
        return {
            ngModule: LocalizationModule,
            providers: [LocalizationService]
        };
    };
    LocalizationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        TranslatePipe,
                        TranslateDirective
                    ],
                    exports: [
                        TranslatePipe,
                        TranslateDirective
                    ]
                },] },
    ];
    /** @nocollapse */
    LocalizationModule.ctorParameters = function () { return []; };
    return LocalizationModule;
}());
function LocalizationModule_tsickle_Closure_declarations() {
    /** @type {?} */
    LocalizationModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    LocalizationModule.ctorParameters;
}
