// @progress/kendo-angular-grid/dist/npm/grouping/group-connection.service.js
"use strict";
var core_1 = require('@angular/core');
var utils_1 = require('../utils');
var Observable_1 = require('rxjs/Observable');
require('rxjs/add/observable/from');
require('rxjs/add/observable/merge');
require('rxjs/add/observable/fromEvent');
require('rxjs/add/operator/take');
require('rxjs/add/operator/takeUntil');
require('rxjs/add/operator/switchMapTo');
require('rxjs/add/operator/switchMap');
require('rxjs/add/operator/skipUntil');
var isOutside = function (target, _a) {
    var pageX = _a.pageX, pageY = _a.pageY;
    var dimensions = target.getBoundingClientRect();
    return !(pageX > dimensions.left && pageX < dimensions.right && pageY > dimensions.top && pageY < dimensions.bottom);
};
var preventDefault = function (e) { return e.originalEvent.preventDefault(); };
var createDropCue = function (container) {
    var cue = document.createElement("div");
    cue.className = "k-grouping-dropclue";
    cue.style.display = "none";
    container.appendChild(cue);
    return {
        hide: function () { return cue.style.display = "none"; },
        position: function (element, e, idx, isLast) {
            var left = element.offsetLeft - cue.offsetWidth;
            if (isLast && e.pageX > element.offsetLeft + element.offsetWidth / 2) {
                left = element.offsetLeft + element.offsetWidth;
                idx += 1;
            }
            cue.style.top = element.offsetTop + "px";
            cue.style.left = left + "px";
            cue.style.display = "";
            return idx;
        },
        remove: function () { return container.removeChild(cue); }
    };
};
/**
 * @hidden
 */
var GroupConnectionService = (function () {
    function GroupConnectionService() {
        this.change = new core_1.EventEmitter();
    }
    GroupConnectionService.prototype.register = function (target) {
        if (utils_1.isPresent(target)) {
            this.target = target;
            this.cue = createDropCue(target);
        }
        return this.change.asObservable();
    };
    GroupConnectionService.prototype.isOutside = function (e) {
        return isOutside(this.target, e);
    };
    GroupConnectionService.prototype.showCue = function (e) {
        var item = this.items.filter(function (x) { return !isOutside(x, e); })[0];
        if (item) {
            var index = this.items.indexOf(item);
            this.currentIndex = this.cue.position(item, e, index, index === this.items.length - 1);
        }
        else {
            this.cue.hide();
            this.currentIndex = 0;
        }
    };
    GroupConnectionService.prototype.registerItems = function (items) {
        this.items = items;
    };
    GroupConnectionService.prototype.add = function (field, idx) {
        if (idx === void 0) { idx = this.currentIndex; }
        if (!utils_1.isNullOrEmptyString(field)) {
            this.change.emit({ field: field, idx: idx });
        }
    };
    GroupConnectionService.prototype.hideCue = function () {
        if (utils_1.isPresent(this.cue)) {
            this.cue.hide();
        }
    };
    GroupConnectionService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    GroupConnectionService.ctorParameters = [];
    return GroupConnectionService;
}());
exports.GroupConnectionService = GroupConnectionService;
var append = function (element) {
    var appended = false;
    return function () {
        if (!appended) {
            document.body.appendChild(element);
            appended = true;
        }
        return element;
    };
};
var createHint = function (column) {
    var hint = document.createElement("div");
    hint.className = "k-header k-drag-clue";
    hint.style.cssText = "display:none;position:absolute;";
    hint.innerHTML = "" + (column.title || column.field);
    var getElement = append(hint);
    return {
        move: function (e) {
            preventDefault(e);
            var element = getElement();
            element.style.top = e.pageY + "px";
            element.style.left = e.pageX + "px";
            element.style.display = "";
            return element;
        },
        remove: function () { return document.body.removeChild(hint); }
    };
};
/**
 * @hidden
 */
var GroupDragService = (function () {
    function GroupDragService(connection, ngzone) {
        this.connection = connection;
        this.ngzone = ngzone;
        this.subscriptions = [];
    }
    GroupDragService.prototype.connect = function (draggables, filter) {
        var _this = this;
        if (filter === void 0) { filter = function () { return true; }; }
        this.ngzone.runOutsideAngular(function () {
            _this.unsubscribe();
            _this.subscriptions = draggables.map(function (draggable) {
                var presses = Observable_1.Observable.from(draggable.kendo.press);
                var drags = Observable_1.Observable.from(draggable.kendo.drag);
                var releases = Observable_1.Observable.from(draggable.kendo.release);
                return presses
                    .filter(filter)
                    .do(preventDefault)
                    .switchMap(function (_a) {
                    var column = _a.column;
                    return drags
                        .do(function (e) { return _this.connection.showCue(e); })
                        .map(createHint(column).move);
                })
                    .switchMap(function (hint) { return releases.take(1).do(function () {
                    hint.remove();
                    _this.connection.hideCue();
                }); })
                    .filter(function (e) { return !_this.connection.isOutside(e); })
                    .subscribe(function () { return _this.connection.add(draggable.column.field); });
            });
        });
    };
    GroupDragService.prototype.unsubscribe = function () {
        (this.subscriptions || []).forEach(function (x) { return x.unsubscribe(); });
        this.subscriptions = [];
    };
    GroupDragService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    GroupDragService.ctorParameters = [
        { type: GroupConnectionService, },
        { type: core_1.NgZone, },
    ];
    return GroupDragService;
}());
exports.GroupDragService = GroupDragService;
